lib/spack/spack/spec_list.py

```diff
@@ -189,7 +190,18 @@ def _expand_matrix_constraints(object, specify=True):
         # Construct a combined spec to test against excludes
         flat_combo = [constraint for list in combo for constraint in list]
         ordered_combo = sorted(flat_combo, key=spec_ordering_key)
+
         test_spec = Spec(' '.join(ordered_combo))
+        # Abstract variants don't have normal satisfaction semantics
+        # Convert all variants to concrete types.
+        # This method is best effort, so all existing variants will be
+        # converted before any error is raised.
+        # Catch exceptions because we want to be able to operate on
+        # abstract specs without needing package information
+        try:
+            spack.variant.substitute_abstract_variants(test_spec)
+        except spack.variant.UnknownVariantError:
+            pass
```

Comment:
This seems the right fix to me. For the record, the semantics intended for `AbstractVariant` is that of a name-value variant when no package information is attached. As such it "could satisfy" every other name-value variant with the same name. The `strict` parameter, iirc was not added on purpose - `AbstractVariant` was not meant to be called with `strict=True` and not adding the variant was ensuring an exception is raised if that ever happened.

There's an old discussion at https://github.com/spack/spack/issues/4135#issuecomment-299514214 on which changes we might do to make the code less "surprising" and naming more appropriate. Don't know if everything still applies.